ledger-api/grpc-definitions/com/daml/ledger/api/v1/transaction_filter.proto

```diff
@@ -12,30 +12,102 @@ option java_outer_classname = "TransactionFilterOuterClass";
 option java_package = "com.daml.ledger.api.v1";
 option csharp_namespace = "Com.Daml.Ledger.Api.V1";
 
-// Used for filtering Transaction and Active Contract Set streams.
-// Determines which on-ledger events will be served to the client.
+
+// Filtering Transaction, TransactionTree, and Active Contract Set streams
+//////////////////////////////////////////////////////////////////////////
+
+// A filter both for filtering create and archive events; and for filtering
+// transaction trees.
 message TransactionFilter {
 
-  // Keys of the map determine which parties' on-ledger transactions are being queried.
-  // Values of the map determine which events are disclosed in the stream per party.
-  // At the minimum, a party needs to set an empty Filters message to receive any events.
   // Each key must be a valid PartyIdString (as described in ``value.proto``).
+  // For transaction trees only the listed parties matter, and all subtrees
+  // whose root is has one of the listed parties as an informee are returned.
+  // Create and and archive events are returned for all contracts whose
+  // stakeholders include at least one of the listed parties and match the
+  // per-party filter.
   // Required
   map<string, Filters> filters_by_party = 1;
 }
 
+// The union of a set of contract filters, or a wildcard.
 message Filters {
 
-  // If not set, no filters will be applied.
+  // If set, then contracts matching any of the ``InclusiveFilters`` match
+  // this filter.
+  // If not set, any contract matches this filter.
   // Optional
   InclusiveFilters inclusive = 1;
 }
 
-// If no internal fields are set, no filters will be applied.
+// A filter for contracts.
 message InclusiveFilters {
 
-  // A collection of templates.
+  // A collection of templates for which the payload will be included in the
+  // create_arguments of CreatedEvents.
   // SHOULD NOT contain duplicates.
-  // Required
+  // Optional
   repeated Identifier template_ids = 1;
+
+  // Collection of filters matching on interfaces and the values that should
+  // be included for events matching the filter.
+  // Optional
+  repeated InterfaceFilter interface_filters = 2;
 }
+
+// This filter matches contracts that implement a specific interface.
+message InterfaceFilter {
+
+  // The interface that a matching contract must implement.
+  // Required
+  Identifier interface_id = 1;
+
+  // Whether to include the canonical interface view on the contract in the
+  // returned CreateEvent.
+  // Use this to access contract data in a uniform manner in your Ledger API
+  // client.
+  // Optional.
+  boolean include_interface_view = 2;
+
+  // Whether to include a create_arguments_blob in the returned CreateEvent.
+  // Use this to access the complete contract data in your Ledger API client
+  // for submitting it as a disclosed contract with future commands.
+  // Optional.
+  boolean include_create_arguments_blob = 3;
```

Comment:
The expectation is that disclosures accept both the decoded form available from `create_arguments` and the blob form from `create_arguments_blob`.

Reply:
This is a good point. @meiersi-da & I were thinking off supporting _both_ the current API of passing a value as well as the blob for disclosed contracts which would mean that you’re not blocked from passing templates but you would never be able to use the blob for template ids. I think for CN that’s perfectly fine so I’d be inclined to go with this first. wdyt @meiersi-da

Reply:
race condition, sounds like we agree :)

Reply:
Hm, this feels a bit strange to me. If we support passing a `Record` value, then why do we need the opaque blob? I thought the idea of the opaque blob is that you can include things like driver metadata or a contract payload signature without having to expose them to the user.

If I renamed this parameter to `boolean include_create_arguments` instead and put the contract arguments into `create_arguments` instead of `create_arguments_blob`, what would I loose?

Reply:
You lose LF version independence: If I subscribe to an interface in LF version 1.15 and there is an implementation in LF version 1.16 my client might not be able to decode this without losing information (if you enable unknown fields you can but I believe we don’t even do that in our own bindings atm) so you wouldn’t be able to submit it.

Reply:
... whereas if you filter by template id, the client always knows about the LF version used for that template, so it can properly decode the argument?

But then what if I share that payload off-band with some other client that doesn't know about the required LF version? Wouldn't it be a better user experience if we said disclosed contracts, no matter how you got them, are an opaque blob?

Reply:
> ... whereas if you filter by template id, the client always knows about the LF version used for that template, so it can properly decode the argument?

exactly

> But then what if I share that payload off-band with some other client that doesn't know about the required LF version? 

I think you have two options here: Either that client uses it via its template. In that case they know about the template and therefore about the LF version. Or they use it via interface. In that case, they use it via the blob.

> Wouldn't it be a better user experience if we said disclosed contracts, no matter how you got them, are an opaque blob?

I think there are two aspects here:
1. Should we accept decoded values + metadata? I think this is actually a nice feature. If you’re working with templates you don’t have to carry around both the blob, the decoded payload (because you want to read the data) and the metadata (which also has useful info like let). Instead you can only deal with the latter two.
2. Should we accept blobs for templates? There my argument is mainly one of implementation complexity. If we focus this change only on interfaces, we can get away with slightly simpler APIs which seems nice. I don’t think there is a strong reason to disallow blobs for templates just from a user pov but we don’t have a case where it would be required.

That said, I don’t feel particularly strongly about this so if accepting blobs for templates is easy (or even only accepting blobs) & we come up with a simple extension to the filter & create events API, I’m happy with that as well. I would however, keep at least the let & contract key hash outside of the blob. Those have value on their own as client side information.